pkg/controller/istiocsr/controller.go

```diff
@@ -0,0 +1,331 @@
+		} else if requeue {
+			return ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
+		}
+
+		if err := r.removeFinalizer(ctx, istiocsr, finalizer); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		r.log.V(1).Info("removed finalizer, cleanup complete", "request", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	// Set finalizers on the istiocsr.openshift.operator.io resource
+	if err := r.addFinalizer(ctx, istiocsr); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update %q istiocsr.openshift.operator.io with finalizers: %w", req.NamespacedName, err)
+	}
+
+	return r.processReconcileRequest(istiocsr, req.NamespacedName)
+
+}
+
+func (r *Reconciler) processReconcileRequest(istiocsr *v1alpha1.IstioCSR, req types.NamespacedName) (res ctrl.Result, err error) {
+	defer func() {
+		// using named return values to handle status update in deferred function
+		errUpdate := r.updateStatus(r.ctx, istiocsr)
+		if errUpdate != nil {
+			errUpdate = fmt.Errorf("failed to update %q status: %w", req, errUpdate)
+			err = utilerrors.NewAggregate([]error{err, errUpdate})
+		}
+	}()
+
+	if err := r.reconcileIstioCSRDeployment(istiocsr); err != nil {
+		r.log.Error(err, "failed to reconcile IstioCSR deployment", "request", req)
+		if IsIrrecoverableError(err) {
+			istiocsr.Status.SetCondition(v1alpha1.Degraded, metav1.ConditionTrue, v1alpha1.ReasonFailed, fmt.Sprintf("reconciliation failed with irrecoverable eror not retrying: %v", err))
+			istiocsr.Status.SetCondition(v1alpha1.Ready, metav1.ConditionFalse, v1alpha1.ReasonReady, "")
+			return ctrl.Result{}, nil
+		} else {
+			istiocsr.Status.SetCondition(v1alpha1.Degraded, metav1.ConditionFalse, v1alpha1.ReasonReady, "")
+			istiocsr.Status.SetCondition(v1alpha1.Ready, metav1.ConditionFalse, v1alpha1.ReasonInProgress, fmt.Sprintf("reconciliation failed, retrying: %v", err))
+			return ctrl.Result{RequeueAfter: defaultRequeueTime}, fmt.Errorf("failed to reconcile %q IstioCSR deployment: %w", req, err)
+		}
+	}
+
+	istiocsr.Status.SetCondition(v1alpha1.Degraded, metav1.ConditionFalse, v1alpha1.ReasonReady, "")
+	istiocsr.Status.SetCondition(v1alpha1.Ready, metav1.ConditionTrue, v1alpha1.ReasonReady, "reconciliation successful")
+	return ctrl.Result{}, nil
+}
+
+// cleanUp handles deletion of istiocsr.openshift.operator.io gracefully.
+func (r *Reconciler) cleanUp(istiocsr *v1alpha1.IstioCSR) (bool, error) {
+	// TODO: For GA, handle cleaning up of resources created for installing istio-csr operand.
+	// This might require a validation webhook to check for usage of service as GRPC endpoint in
+	// any of OpenShift Service Mesh or Istiod deployments to avoid disruptions across cluster.
+	r.eventRecorder.Event(istiocsr, corev1.EventTypeWarning, "RemoveDeployment", "%s istiocsr marked for deletion, remove reference in istiod deployment and remove all resources created for istiocsr deployment")
```

Comment:
```suggestion
	r.eventRecorder.Eventf(istiocsr, corev1.EventTypeWarning, "RemoveDeployment", "%s istiocsr marked for deletion, remove reference in istiod deployment and remove all resources created for istiocsr deployment")
```

without Eventf the "%s" remains unescaped.

Reply:
suggestion incorporated.